Add miri to rustc docs.rs page

InfRandomness opened the pull request:
This adds miri to https://doc.rust-lang.org/nightly/nightly-rustc/

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

InfRandomness commented:
I couldn't figure out a way to generate [doc.rust-lang.org/nightly/nightly-rustc](https://doc.rust-lang.org/nightly/nightly-rustc/)
locally
I tried running `/.x.py doc` but it failed, so checks might fail here too

InfRandomness commented:
The pipeline doesn't seem to have failed, but since it failed on my side I don't think the same commands are being ran

jyn514 commented:
> The pipeline doesn't seem to have failed, but since it failed on my side I don't think the same commands are being ran

Correct, documentation is only built in bors try builds. You'll need to get x.py doc working locally before this is being merged. Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference) or allow the bootstrap warnings for miri in doc.rs. You should be able to replicate the warnings out of tree with `RUSTFLAGS="-Drustc::all -Zunstable-options" cargo doc`.

jyn514 commented:
> Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference

ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter: https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710

I think `tool.rs` does something similar, you could copy from there.

InfRandomness commented:
> > Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference
> 
> ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter:
> 
> https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710
> 
> I think `tool.rs` does something similar, you could copy from there.

You mean, if there is a tool_doc!() call in tool.rs ? 
I didn't find any but I'll ripgrep and see what I can find

InfRandomness commented:
> > Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference
> 
> ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter:
> 
> https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710
> 
> I think `tool.rs` does something similar, you could copy from there.

tool.rs effectively seems to have https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/tool.rs#L767
but I don't really know what to do to tell the code to generate docs for that, now that i've reverted my original diff

jyn514 commented:
@InfRandomness the `tool.rs` file is for building miri, not documenting it. You need to change `doc.rs` to document miri.

I don't understand the rest of your question, sorry.

InfRandomness commented:
> @InfRandomness the `tool.rs` file is for building miri, not documenting it. You need to change `doc.rs` to document miri.
> 
> I don't understand the rest of your question, sorry.

did you mean that I need to end up with something similar to : 
```
tool_doc!(Miri, "miri", "src/tools/miri", ["miri"], submodule=true);
```
?

InfRandomness commented on src/bootstrap/doc.rs:
```diff
@@ -729,15 +735,16 @@ macro_rules! tool_doc {
 }
 
-tool_doc!(Rustdoc, "rustdoc-tool", "src/tools/rustdoc", ["rustdoc", "rustdoc-json-types"]);
+tool_doc!(Rustdoc, "rustdoc-tool", "src/tools/rustdoc", ["rustdoc", "rustdoc-json-types"], true);
```
Something akin to [`499ca27` (#97773)](https://github.com/rust-lang/rust/pull/97773/commits/499ca27c974b2c70db1b3a26d8b5ae8bec4607bb) ?

jyn514 replied:
Yup, exactly :)

jyn514 reviewed (approved):
Thanks!

bors commented:
:pushpin: Commit 499ca27c974b2c70db1b3a26d8b5ae8bec4607bb has been approved by `jyn514`

<!-- @bors r=jyn514 499ca27c974b2c70db1b3a26d8b5ae8bec4607bb -->
<!-- homu: {"type":"Approved","sha":"499ca27c974b2c70db1b3a26d8b5ae8bec4607bb","approver":"jyn514"} -->

jyn514 commented:
Actually, @InfRandomness can you confirm that `x.py doc src/tools/miri --stage 0` works for you after the latest commit?

InfRandomness commented:
> Actually, @InfRandomness can you confirm that `x.py doc src/tools/miri --stage 0` works for you locally after the latest commit?

hm, nope, for some reason it panicked with the command you provided

jyn514 commented:
> I do seem to be having a couple of warnings coming from rustdocs and I would like to fix them beforehand

:+1: 

@bors r-

RalfJung commented on src/bootstrap/doc.rs:
```diff
@@ -729,14 +735,22 @@ macro_rules! tool_doc {
     "rustfmt-nightly",
     "src/tools/rustfmt",
     ["rustfmt-nightly", "rustfmt-config_proc_macro"],
+    in_tree = true
```
FWIW I think `in_tree: true` might be more Rust-style syntax here.

RalfJung replied:
> in_tree: true feels more like when you are constructing a structure,

Yes, exactly. That is an argument in favor of `:`, isn't it?

`in_tree = true` looks like a `let`-binding, but that would mean the name is *our* choice, which it is not.

jyn514 replied:
This is consistent with the macro in `tool.rs`, I'd prefer to keep it as-is.
https://github.com/rust-lang/rust/blob/c62c482b82f4ea5d3d6b87d5f85b87f2fb0dc4ab/src/bootstrap/tool.rs#L693

RalfJung replied:
Ah, fair. Sorry for the noise.

jyn514 replied:
> so what do I do ? roll back

I'd appreciate that, thanks. You'll need to rebase this PR over the miri update in any case.

InfRandomness replied:
> > so what do I do ? roll back
> 
> I'd appreciate that, thanks. You'll need to rebase this PR over the miri update in any case.

For some reason, git refuses to push the rebased branch

```
  1 src/bootstrap/ git … push -v --force fork refs/heads/miri-docs\:refs/heads/miri-docs
Pushing to github.com:InfRandomness/rust.git
Running pre-push script '/home/infrandomness/Documents/Dev/rust/x.py test tidy --bless'
Updating only changed submodules
Updating submodule src/tools/miri
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0        
fatal: 'fork' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: Fetched in submodule path 'src/tools/miri', but it did not contain 3361eabf3882578207a483f2cee631646e80eabc. Direct fetching of that commit failed.
Failed updating submodule. This is probably due to uncommitted local changes.
Either stash the changes by running "git stash" within the submodule's
directory, reset them by running "git reset --hard", or commit them.
To reset all submodules' changes run "git submodule foreach --recursive git reset --hard".
Build completed unsuccessfully in 0:00:00
error: failed to push some refs to 'github.com:InfRandomness/rust.git'
```

jyn514 replied:
This is a bug in git - try running `git submodule deinit -f src/tools/miri` and seeing if it fixes the issue.
https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/More.20submodule.20problems

InfRandomness replied:
> `git submodule deinit -f src/tools/miri`

It worked but I had to remove the pre-push hook that invoked x.py because it reinitialized the submodule everytime

InfRandomness commented:
This MR is on hold until the MIRI submodule gets updated

RalfJung commented:
The Miri update landed. :)

InfRandomness commented:
Ok this is now ready to be merged

jyn514 commented:
> Ok this is now ready to be merged

@InfRandomness did you run `x doc --stage 0 src/tools/miri` locally to make sure it works?

InfRandomness commented:
> > Ok this is now ready to be merged
> 
> @InfRandomness did you run `x doc --stage 0 src/tools/miri` locally to make sure it works?

I tried and it fails with the submodule git bug

InfRandomness commented:
It started working again after cloning back up the repository, and now I have another warning I did not have before ._.

RalfJung commented:
Is there a way to generate these warnings in a clone of the Miri repo? Can we have a check in Miri CI which ensures that the rustdoc check passes?

RalfJung commented:
Ah, this seems to work (slightly different than the command @jyn514 shared above):
```
RUSTDOCFLAGS="-Drustdoc::all -Zunstable-options" cargo doc
```

@InfRandomness maybe a good next step would be to add that to Miri's CI.

InfRandomness commented:
> Ah, this seems to work (slightly different than the command @jyn514 shared above):
> 
> ```
> 
> RUSTDOCFLAGS="-Drustdoc::all -Zunstable-options" cargo doc
> 
> ```
> 
> 
> 
> @InfRandomness maybe a good next step would be to add that to Miri's CI.

Yep, definitely, it would prevent this from happening again

InfRandomness commented:
This MR will be able to land once https://github.com/rust-lang/miri/pull/2222 lands, and once the miri submodule gets updated

jyn514 commented:
@InfRandomness what's the status of this PR? Is it waiting on the submodule update? If not, can you rebase over the submodule update and confirm the warning has been fixed?

InfRandomness commented:
> @InfRandomness what's the status of this PR? Is it waiting on the submodule update? If not, can you rebase over the submodule update and confirm the warning has been fixed?

It is waiting on the submodule update

jyn514 commented:
You can also update the submodule in this PR :) it's ok to do both at once.

InfRandomness commented:
`./x.py doc --stage 0 src/tools/miri` seems to be running fine after updating both the submodule and rebasing on top of master

jyn514 commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 7932f823b0fc912df339e28ca0ffac3a7b145e15 has been approved by `jyn514`

<!-- @bors r=jyn514 7932f823b0fc912df339e28ca0ffac3a7b145e15 -->
<!-- homu: {"type":"Approved","sha":"7932f823b0fc912df339e28ca0ffac3a7b145e15","approver":"jyn514"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6941132216?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7932f823b0fc912df339e28ca0ffac3a7b145e15/checks/6941132216/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 43c47db0b04627dbd0e16a1e7cad14a4a5a60d3a and cbbc132be88c566b69c8aa91765d288c287f84eb
Submodules were updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
Pass got exit status: 101

actual output differed from expected tests/pass/btreemap.stderr
Diff < left / right > :
>error: internal compiler error: /checkout/compiler/rustc_const_eval/src/interpret/place.rs:LL:CC: write_immediate_to_mplace: invalid ScalarPair layout: TyAndLayout {
>                                    ty: *mut alloc::collections::btree::node::LeafNode<Foo, ()>,
>                                    layout: Layout {
>                                        fields: Primitive,
>                                        variants: Single {
>                                        },
>                                        abi: Scalar(
>                                            Initialized {
>                                                value: Pointer,
>                                                value: Pointer,
>                                                valid_range: 0..=18446744073709551615,
>                                            },
>                                        ),
>                                        largest_niche: None,
>                                        align: AbiAndPrefAlign {
>                                            abi: Align(8 bytes),
>                                            pref: Align(8 bytes),
>                                        size: Size(8 bytes),
>                                    },
>                                }
>  --> RUSTLIB/alloc/src/collections/btree/node.rs:LL:CC
>  --> RUSTLIB/alloc/src/collections/btree/node.rs:LL:CC
>   |
>LL |             LeafNode::init(leaf.as_mut_ptr());
>
>thread 'rustc' panicked at 'Box<dyn Any>', /checkout/compiler/rustc_errors/src/lib.rs:LL:CC
>stack backtrace:
>stack backtrace:
>   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hfdc6954014fcf611
>   1: core::fmt::write::h16d774585cf0317e
>   2: std::io::Write::write_fmt::h60b53a2506f43cae
>   3: std::panicking::default_hook::{{closure}}::hba167fe557504bf3
>   4: std::panicking::default_hook::h6d3db90619ad2c86
>   5: rustc_driver[977c616eac25824]::DEFAULT_HOOK::{closure#0}::{closure#0}
>   6: std::panicking::rust_panic_with_hook::h7e209aca36b08087
>   7: std[df7aec823036d165]::panicking::begin_panic::{closure#0}
>   8: std[df7aec823036d165]::sys_common::backtrace::__rust_end_short_backtrace
>   9: std[df7aec823036d165]::panicking::begin_panic
>  10: std[df7aec823036d165]::panic::panic_any
>  11: <rustc_errors[a2363adab38fd74a]::HandlerInner>::span_bug
>  12: <rustc_errors[a2363adab38fd74a]::Handler>::span_bug
>  13: rustc_middle[fc654091caa9cc97]::ty::context::tls::with_opt
>  14: rustc_middle[fc654091caa9cc97]::util::bug::opt_span_bug_fmt
>  15: rustc_middle[fc654091caa9cc97]::util::bug::span_bug_fmt
>  16: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::write_immediate_to_mplace_no_validate
>  17: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::write_immediate_no_validate
>  18: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::copy_op_no_validate
>  19: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::copy_op_transmute
>  20: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::pop_stack_frame
>  21: <rustc_const_eval[7f1c0917a782fb4e]::interpret::eval_context::InterpCx<miri[1ac70ecead0b1656]::machine::Evaluator>>::terminator
>  22: miri[1ac70ecead0b1656]::eval::eval_entry
>  23: <rustc_interface[3927467015402c63]::passes::QueryContext>::enter
>  24: <miri[77eb59cf771f1aea]::MiriCompilerCalls as rustc_driver[977c616eac25824]::Callbacks>::after_analysis
>  25: <rustc_interface[3927467015402c63]::interface::Compiler>::enter
>  26: rustc_span[ba014c4bfd7e5620]::with_source_map
>  27: <scoped_tls[fd34f0afc4d58711]::ScopedKey<rustc_span[ba014c4bfd7e5620]::SessionGlobals>>::set
>  28: std[df7aec823036d165]::sys_common::backtrace::__rust_begin_short_backtrace
>  29: <<std[df7aec823036d165]::thread::Builder>::spawn_unchecked_<rustc_interface[3927467015402c63]::util::run_in_thread_pool_with_globals<rustc_interface[3927467015402c63]::interface::run_compiler<core[2b2671207cb872ad]::result::Result<(), rustc_errors[a2363adab38fd74a]::ErrorGuaranteed>, rustc_driver[977c616eac25824]::run_compiler::{closure#1}>::{closure#0}, core[2b2671207cb872ad]::result::Result<(), rustc_errors[a2363adab38fd74a]::ErrorGuaranteed>>::{closure#0}, core[2b2671207cb872ad]::result::Result<(), rustc_errors[a2363adab38fd74a]::ErrorGuaranteed>>::{closure#1} as core[2b2671207cb872ad]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
>  30: std::sys::PLATFORM::thread::Thread::new::thread_start::h5ab75bff0890bb5f
>  31: start_thread
>  32: clone
>  33: <unknown>
>note: the compiler unexpectedly panicked. this is a bug.
>
>note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
>
>
>note: rustc 1.63.0-nightly (cbbc132be 2022-06-17) running on x86_64-unknown-linux-gnu
>
>note: compiler flags: -Z ui-testing -Z miri-strict-provenance
>query stack during panic:
>end of query stack
>error: aborting due to previous error
>
>
>



There were 1 unmatched diagnostics that occurred outside the testfile and had not pattern
    Ice: /checkout/compiler/rustc_const_eval/src/interpret/place.rs:771:68: write_immediate_to_mplace: invalid ScalarPair layout: TyAndLayout {
    ty: *mut alloc::collections::btree::node::LeafNode<Foo, ()>,
    layout: Layout {
        fields: Primitive,
        variants: Single {
        },
        abi: Scalar(
            Initialized {
                value: Pointer,
                value: Pointer,
                valid_range: 0..=18446744073709551615,
            },
        ),
        largest_niche: None,
        align: AbiAndPrefAlign {
            abi: Align(8 bytes),
            pref: Align(8 bytes),
        size: Size(8 bytes),
    },
}

---
This PR updated 'src/tools/miri', verifying if status is 'test-pass'...

We detected that this PR updated 'miri', but its tests failed.

If you do intend to update 'miri', please check the error messages above and
commit another update.

If you do NOT intend to update 'miri', please ensure you did not accidentally
change the submodule at 'src/tools/miri'. You may ask your reviewer for the
Build completed unsuccessfully in 0:00:00
```

</details>

InfRandomness commented:
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6941132216?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7932f823b0fc912df339e28ca0ffac3a7b145e15/checks/6941132216/logs)
> Click to see the possible cause of the failure (guessed by this bot)

Huh

jyn514 commented:
@bors r-

@RalfJung this looks like a failure from the submodule update, not sure if you recognize it

InfRandomness commented:
> @bors r-
> 
> @RalfJung this looks like a failure from the submodule update, not sure if you recognize it

I see that it said 
`We detected that this PR updated 'miri', but its tests failed.`
but I'm not sure how to reproduce that on my local machine,
do you know ?

jyn514 commented:
>      Running unittests src/lib.rs (build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/miri-b55e52d0cc2c0e6d)

you can probably reproduce with `x test --stage 2 src/tools/miri`. Not sure whether --stage 2 is necessary or not, --stage 1 might work.

RalfJung commented:
> this looks like a failure from the submodule update, not sure if you recognize it

Yes, that's https://github.com/rust-lang/rust/issues/98107

RalfJung commented:
However, this PR shouldn't have to update the submodule. Yes, Miri fails to build on master, but that happens regularly so the documentation generation code here has to be able to handle that situation.

RalfJung commented on src/bootstrap/doc.rs:
```diff
@@ -707,7 +713,7 @@ macro_rules! tool_doc {
-                    SourceType::InTree,
+                    source_type,
                     &[],
                 );
```
If this command fails for Miri, that failure must be ignored. Miri failing to build is not a bug, it is expected. That's what the entire toolstate stuff is about.

jyn514 commented:
@RalfJung this has to update the submodule to pull in the documentation fix, not so it gets miri building.

Oh I see what you're saying though - if miri is broken, the docs will be broken too and we have to be able to handle that. @InfRandomness I think the right way to do that is to change https://github.com/rust-lang/rust/blob/7932f823b0fc912df339e28ca0ffac3a7b145e15/src/bootstrap/doc.rs#L732 to use `try_run` if `$in_tree` is true, and maybe print a warning if the command fails.

RalfJung commented:
> this has to update the submodule to pull in the documentation fix, not so it gets miri building.

The documentation fix is already in the submodule in rustc master. The submodule update in this PR currently does not contain any doc fixes.

RalfJung commented:
> to use try_run if $in_tree is true, and maybe print a warning if the command fails.

You mean `false`? Still `in_tree` seems like too innocent of a name to also control ignoring build errors...

jyn514 commented:
:shrug: there are only ~3 tools in tree that use submodules, I don't think it's going to cause problems in practice.

bors commented:
:umbrella: The latest upstream changes (presumably #98237) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

InfRandomness commented:
I don't know how to resolve a conflict with a submodule,
regular 'git add' doesn't seem to work...
anyone knows?

RalfJung commented:
I'd recommend to first remove the submodule update from this commit, and then rebase it.

InfRandomness commented:
> I'd recommend to first remove the submodule update from this commit, and then rebase it.

With `git restore`?
that doesn't seem to work...

RalfJung commented:
The submodule update is already a separate commit. So just `git reset --hard HEAD^` should do it. Then fix the submodules (`git submodule update --init --recursive`) and then rebase.

InfRandomness commented:
> The submodule update is already a separate commit. So just `git reset --hard HEAD^` should do it. Then fix the submodules (`git submodule update --init --recursive`) and then rebase.

```
LANG=en_US git submodule update --init --recursive
fatal: 'fork' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: Fetched in submodule path 'src/tools/cargo', but it did not contain 4d92f07f34ba7fb7d7f207564942508f46c225d3. Direct fetching of that commit failed.
```

RalfJung commented:
Looks like you got your checkout into a broken state. I remember seeing some weird behavior with submodule+rebase in the past (git is definitely terrible at handling this), but I have no idea what I did to get out of it again. Probably random bashing against various git commands until it gave in...

You also had bad luck; there was some trouble with the cargo and rust-analyzer submodules right around when you tried that (that is probably the reason of that cargo-related error).

What you want to do is get back to commit d23264b9df5bd9d10b5553956b9ae523b1125055, somehow. If it is still rebasing, try `git rebase --abort`. Then `git reset --hard d23264b9df5bd9d10b5553956b9ae523b1125055`. Then `git rebase origin/master`. And then `git submodule update --init --recursive`. Then hopefully `git status` will be clean and you can push to this PR's branch.

InfRandomness commented:
I tried to git checkout the mentioned commit 
```
[infrandomness@fedora rust]$ git status
HEAD detached at d23264b9df5
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   Cargo.lock
	modified:   src/tools/cargo (new commits)
	modified:   src/tools/miri (new commits)
	modified:   src/tools/rust-analyzer (new commits, modified content)

no changes added to commit (use "git add" and/or "git commit -a")
[infrandomness@fedora rust]$ git rebase --abort
fatal: No rebase in progress?
```

InfRandomness commented:
Ok I've had enough of this git bullcrap

I'm scraping all of this with my fork, recreating a new fork and a new PR.

InfRandomness commented:
Nvm I won't even bother opening a new PR
I'm done with this.

RalfJung commented:
Ah, one submodule has "modified content". Not sure if you edited it or some script changed a file in there, but I don't know of a good way to tell git to reset all submodule *contents*, you have to go to each affected submodule and do it by hand. :(

I re-submitted this as https://github.com/rust-lang/rust/pull/98714.
